feat: support live reload with tilt.dev

marcja opened the pull request:
Fixes #38 

Adds support for live reload per best practices recommended at tilt.dev -- see [NodeJS](https://docs.tilt.dev/example_nodejs.html) and [Java](https://docs.tilt.dev/example_java.html). Bringing up the app is the same as before -- assuming you've completed your platform-specific configuration of ingress-nginx, just change into the `callmeback` folder and `tilt up`.

With this solution, you can edit a JS/CSS file and (almost) immediately see your changes reflected in your browser without refreshing. Similarly for Java files.

- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR
- [ ] Feature has been fleshed out in design document

kebrahim commented on callmeback/database/docker-entrypoint-initdb.d/init.js:
```diff
@@ -0,0 +1,9 @@
+// initialization goes here
+db = db.getSiblingDB('nyst');
+db.reservation.insertOne({
```
what is the point of this? is this a test?

marcja replied:
Removed. It's example code for how you can initialize a brand new database .

kebrahim commented on callmeback/database/k8s.yml:
```diff
@@ -0,0 +1,63 @@
+  labels:
+    app.kubernetes.io/name: database
+    app.kubernetes.io/component: database
+    app.kubernetes.io/part-of: backend
```
more of a k8s question, but what does the fact that this component is "part-of" the backend component imply? i.e. why are the backend & frontend "part-of" `callmeback`, but the database is "part-of" the `backend`?

marcja replied:
It's basically "for show". While this is recommending style, it's basically only used by k8s orchestration apps (for example, that might want to visualize k8s resources as a graph). It doesn't do anything with the tools we do.

itssumertime commented on callmeback/frontend/Dockerfile:
What is this and the below line repeating the item twice for "ADD" do?

marcja replied:
Every line in a Dockerfile is a _layer_. You can think if this also like a checkpoint. Basically, Docker will try to preserve/cache everything up to the first change it detects. So, if `package.json` hasn't but `yarn.lock` has, Docker will use what it has cached up to the `ADD yarn.lock yarn.lock` line and rebuild the image from there. This can be a major performance optimization.

itssumertime commented on callmeback/Tiltfile:
```diff
@@ -1,11 +1,5 @@
-k8s_yaml(kustomize('kubernetes'))
+include('./database/Tiltfile')
```
Now that there are individual Tiltfiles all referenced here, will "tilt up" still work the same way from this file? If there are any changes, could you please also update the README steps?

marcja replied:
Yes it will!

kebrahim commented:
please update the README to remove this "Known issue" - https://github.com/google-org/new-project-ny/tree/master/callmeback#known-issues

marcja commented:
So the build-and-test failure is actually because our GitHub action needs to be updated. It assumes that we can call `./gradlew test` in the `callmeback` directory, but this is no longer true. We'll need to setup/call `./gradlew test` in the `callmeback/backend` directory and `yarn test`

kebrahim commented on callmeback/Tiltfile:
```diff
@@ -1,11 +1,5 @@
-k8s_yaml(kustomize('kubernetes'))
```
not sure if this is an issue or expected, but when i 'tilt down', i get this message:

```
Falling back to `kubectl kustomize` since `kustomize` was not found in PATH
```

marcja replied:
That is expected. That also appears on `master`. It just means that all of us are using the version of kustomize built into kubectl (which is an older version) than the latest version which can be separately installed.

kebrahim commented on callmeback/database/Tiltfile:
```diff
@@ -0,0 +1,5 @@
+# and an entry point that starts a server.
+
+k8s_yaml(kustomize('.'))
+k8s_resource('database')
```
i'm having an issue using the Mongo shell - https://github.com/google-org/new-project-ny/tree/master/callmeback#access-mongo-shell

i can access it fine, but when i get to the last step & enter the `db.reservation.find()` command, nothing gets output, even though there is definitely data in the database (i see it in the HAL explorer).

one thing to note - when i run the `kubectl get pods` command, the name of the mongo pod is now `database-0`, when it used to be called `cmb-mongo-xxxx`. not sure if that matters

marcja replied:
Did you switch into the correct database? From the mongo shell, first do `use nyst`. You can see a list of databases via `show dbs`.

kebrahim replied:
that worked! mind updating that mongo shell documentation to reflect the new database name?

marcja replied:
Ack.

marcja commented on callmeback/backend/build.gradle:
```diff
@@ -8,22 +8,29 @@ group = 'org.google'
-  testCompile 'de.flapdoodle.embed:de.flapdoodle.embed.mongo:2.2.0'
+  implementation 'org.springframework.data:spring-data-rest-hal-explorer'
+  implementation group: 'org.hibernate.common', name: 'hibernate-commons-annotations', version: '5.1.0.Final'
+  testImplementation 'de.flapdoodle.embed:de.flapdoodle.embed.mongo:2.2.0'
```
Ack. Won't be able to look into this today.

marcja replied:
Fixed.

tariqajyusuf commented on callmeback/backend/Tiltfile:
```diff
@@ -0,0 +1,22 @@
+    sync('./build/jar/META-INF', '/app/META-INF'),
+    sync('./build/jar/BOOT-INF/classes', '/app'),
+  ],
+  entrypoint = 'find . | entr -r java -noverify -cp .:./lib/* org.google.callmeback.CallMeBackApp')
```
It seems like what you're trying to do here is restart things only if files change. I believe Tilt already does this and caches old builds (so if you undo a code change the restart is almost instant). Recommend deleting this (and the subsequent package import in `backend/Dockerfile`) unless there's a compelling reason to not use the default behavior.

marcja replied:
So this is verbatim from the recommended approach at tilt.dev. See [Java](https://docs.tilt.dev/example_java.html). This isn't my code, it's Tilt's.

marcja replied:
So it looks like tilt.dev updated their recommended setup since I created my test project. The restart process extension wasn't documented when I started this. Will update.

marcja replied:
Well, extensions don't currently seem to be compatible with `include` directives, so deferring this for now.

tariqajyusuf commented on callmeback/backend/src/main/java/org/google/callmeback/CallMeBackApp.java:
```diff
@@ -11,12 +10,6 @@
 @EnableScheduling
 public class CallMeBackApp {
   public static void main(String[] args) {
-    Properties props = new Properties();
```
Is there a reason you removed the dynamic configuration here? It's a useful component to have implemented.

marcja replied:
Only that it wasn't being used for anything AFAICT.

marcja replied:
Moved this configuration logic into a Spring profile.

tariqajyusuf commented on callmeback/database/k8s.yml:
```diff
@@ -0,0 +1,63 @@
+      - name: database
+        image: mongo:latest
+        env:
+          - name: MONGO_INITDB_ROOT_USERNAME
```
The security part of me want to have the comment back of "you shouldn't do this in practice" in a README or similar file.

marcja replied:
Yes, although the way we were using kustomize to "hide" the secret in plain text in a companion yaml file wasn't really more secure either. The better way is to store the secret in GitHub Secrets, encrypt the secret, or have the secret be a configuration that isn't checked in.

marcja commented:
Addressed most review comments. All that remains is looking into the backend test failure and updating the GitHub Actions (aka presubmit) to accept the new build/test structure.

marcja commented:
Backend tests now pass. PTAL.

I will do the GitHub Actions work in a separate PR.

marcja commented:
Phew. The GitHub Actions are now working. PTAL. This should be ready for merge.
